web-app/HeimScripts/heatmap/run.R

```diff
@@ -393,3 +484,40 @@ addClusteringOutput <- function(jsn, measurements) {
   )
   return(jsn)
 }
+
+### Limma part. It will be moved into a separate
+# module when we get R scripts sourcing to work here.
+
+getDEgenes <- function(df) {
+  measurements    <- getMeasurements(df)
+  design          <- getDesign(measurements)
+  contrast.matrix <- makeContrasts( S1-S2, levels = design )
+  fit             <- lmFit(measurements, design)
+  fit             <- contrasts.fit(fit, contrast.matrix)
+  fit             <- eBayes(fit)
+  contr           <- 1  #  We need a vector, not a df, so we'll do [, contr] on all stats
+  top.fit         <- data.frame (
+    logfold = fit$coefficients[, contr],
+    ttest   = fit$t[, contr],
+    pval    = fit$p.value[, contr],
+    adjpval = p.adjust(
+      p      = fit$p.value[, contr],
+      method ='fdr'),
+    bval    = fit$lods[, contr]
+  )
+  assign("t_fit", top.fit, envir = .GlobalEnv)
+  cbind(df[,1:2], top.fit)
+}
+
+getDesign <- function(measurements) {
+  subset1Length <- getSubset1Length(measurements)
+  classVectorS1 <- c(rep(1, subset1Length), rep(2, ncol(measurements) - subset1Length ))
+  classVectorS2 <- rev(classVectorS1)
+  cbind(S1=classVectorS1, S2=classVectorS2)
```

Comment:
Is it needed to have the things of S1 first and S2 second or can it be mixed?
Otherwise grabbing the subset number from the column names might be more robust than: first sorting, then taking element 1:length(subset S1) to get elements for S1

Could be done for example like this (but there are many ways to do it, e.g. using gsub or strsplit)
this will get "s1" or "s2" 
 subsets_match_indices <- regexpr("s[[:digit:]]+$", column_names)
  subsets_matches <-    regmatches(x = column_names, m = subsets_match_indices)

replacing s by empty string, converting to integer:
subsetNrs<-as.integer(gsub("s","s",subsets_matches))